### PR TITLE
Not only querying, but also modifying system parameters

### PR DESCRIPTION
**Pull request by yuanqing-wang**
- [x] Modifying Parameters in System

### CODE REVIEW

**Comment by yuanqing-wang** [comment]
cross-ref: https://github.com/openmm/openmm-cookbook/issues/12

File: notebooks/cookbook/Modifying Parameters in an OpenMM System.ipynb [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+   "cell_type": "markdown",
+   "id": "5cc09957",
+   "metadata": {},
+   "source": [
+    "## Modifying Parameters in System\n",
+    "\n",
+    "Some machine learning force field construction schemes, such as [Espaloma](https://pubs.rsc.org/en/content/articlehtml/2022/sc/d2sc02739a), would output parameters which can be picked up by the parameter-holding `System` object of OpenMM, which could be further used in molecular dynamics (MD) simulations just like a vanilla `System`.\n",
+    "In this tutorial, we show how you can modify the parameters of `System` object so you can write your own machine learning force field API.\n",
+    "\n",
+    "This example is inspired by the [Espaloma deployment](https://github.com/choderalab/espaloma/blob/master/espaloma/graphs/deploy.py) implementation."
```
Comment (peastman) [inline notebooks/cookbook/Modifying Parameters in an OpenMM System.ipynb:13]: This description is overly specific.  Nothing in this notebook is really specific to machine learning.  You're just showing how to examine and change the parameters in a System.  How about something like this?

> Sometimes it is useful to examine the parameters in a System.  For example, you might want to check what parameters a ForceField has assigned to a particular bond or angle.  In other cases you may want to modify those parameters to make a bond behave differently from what the ForceField assigned.

File: notebooks/cookbook/Modifying Parameters in an OpenMM System.ipynb [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+   "cell_type": "code",
+   "execution_count": 27,
+   "id": "cbdc43ec",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "for idx in range(harmonic_bond_force.getNumBonds()):\n",
+    "    idx0, idx1, r0, k = harmonic_bond_force.getBondParameters(idx)\n",
+    "    if idx0 == 1 and idx1 == 0:\n",
+    "        new_r0 = unit.Quantity(2666, unit.nanometer)\n",
```
Comment (peastman) [inline notebooks/cookbook/Modifying Parameters in an OpenMM System.ipynb:248]: The cleaner way of assigning units is

```python
new_r0 = 2666*unit.nanometer
```

**Review by peastman — COMMENTED** [review]
These notebooks have a lot of overlap with existing ones.  There's already a cookbook entry on [inspecting system parameters](https://openmm.github.io/openmm-cookbook/latest/notebooks/cookbook/Querying%20Charges%20and%20Other%20Parameters.html).  Is there a good reason to have two on the same thing?  The other notebook is esoteric and sort of out of scope: using OpenMM to validate your own MD code that you've written separately.  Mostly that isn't about OpenMM.  The only part that's really about OpenMM is the discussion of how to query energy components, but we already have [a cookbook entry](https://openmm.github.io/openmm-cookbook/latest/notebooks/cookbook/Analyzing%20Energy%20Contributions.html) on that.

**Comment by yuanqing-wang** [comment]
@peastman how about I get rid of the separate "modify the parameters" one and simply append the existing "query the parameters" tutorial with ways to modify them in-place.

For the other one, I have seen so many people trying to using OpenMM to benchmark their own MM implementations. I feel like the community would benefit from a tutorial.

**Comment by yuanqing-wang** [comment]
see https://github.com/openmm/openmm-cookbook/pull/18/commits/38035e44cac7142561dfa1220288efbec358226b where I combined the modifying parameter tutorial with the querying parameter tutorial

**Comment by yuanqing-wang** [comment]
would love your feedback! @peastman @sef43 

Perhaps the second item is more suitable for a tutorial than a cookbook?

File: notebooks/cookbook/Querying and Modifying Charges and Other Parameters.ipynb [hunk 3]
```diff
@@ -5,7 +5,7 @@
    "id": "3a4fdeed",
    "metadata": {},
    "source": [
-    "## Querying Charges and Other Parameters\n",
+    "## Querying and Modifyinig Charges and Other Parameters\n",
```
Comment (sef43) [inline notebooks/cookbook/Querying and Modifying Charges and Other Parameters.ipynb:8]: Modifying typo

**Review by sef43 — CHANGES_REQUESTED** [review]

**Comment by sef43** [comment]
I think the setting different parameters is a good addition to the existing cookbook entry!

I think as you have suggested the notebook for "Use OpenMM to Verify Your MM Functional Form Implementation" should be a separate tutorial. Can you remove it from this PR and make a new PR adding that as a tutorial, we can then review it properly in the new PR.

**Comment by yuanqing-wang** [comment]
@sef43 I've made the changes as requested. Will follow up with another PR for the separate notebook.

**Comment by yuanqing-wang** [comment]
Separated PR here: https://github.com/openmm/openmm-cookbook/pull/22

File: notebooks/cookbook/Querying and Modifying Charges and Other Parameters.ipynb [hunk 4]
```diff
@@ -98,12 +98,81 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "id": "0dc0db66",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "k_ = 2666 * unit.kilojoule / ( unit.nanometer ** 2 * unit.mole )"
```
Comment (peastman) [inline notebooks/cookbook/Querying and Modifying Charges and Other Parameters.ipynb:126]: This can be written more simply as `unit.kilojoules_per_mole/unit.nanometer**2`.  Also, since this value only ever appears once, we can just put it directly in the call to `setBondedParameters()` instead of defining a separate variable for it.

File: notebooks/cookbook/Querying and Modifying Charges and Other Parameters.ipynb [hunk 5]
```diff
@@ -98,12 +98,81 @@
...
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Particles (1, 0), length = 0.101 nm, k = 2666.0 kJ/(nm**2 mol)\n"
+     ]
+    }
+   ],
+   "source": [
+    "bonded = [f for f in system.getForces() if isinstance(f, HarmonicBondForce)][0]\n",
```
Comment (peastman) [inline notebooks/cookbook/Querying and Modifying Charges and Other Parameters.ipynb:165]: We already looked up the HarmonicBondForce.  You don't need to do it again.

File: notebooks/cookbook/Querying and Modifying Charges and Other Parameters.ipynb [hunk 6]
```diff
@@ -98,12 +98,81 @@
...
+    "for i in range(bonded.getNumBonds()):\n",
+    "    particle1, particle2, length, k = bonded.getBondParameters(i)\n",
+    "    if particle1 == 1 and particle2 == 0:\n",
```
Comment (peastman) [inline notebooks/cookbook/Querying and Modifying Charges and Other Parameters.ipynb:168]: Let's make the condition the same as before: `particle1 == 0 or particle2 == 0`.  That way it will print out the same list of bonds, and they can see that only one has changed.

**Review by peastman — COMMENTED** [review]
Looks good!  I made a few very minor suggestions.

**Comment by peastman** [comment]
I just realized it's missing one very important part.  Let's add the following at the end.

> Modifying Force objects will affect any new Simulations or Contexts you create, but it will have no effect on ones that already exist.  If you want your modifications to apply to an existing Simulation, you can copy them over by calling `bonded.updateParametersInContext(simulation.context)`.

**Comment by yuanqing-wang** [comment]
@peastman I've made the changes per your suggestions. I've also added the code to modify existing `Context`.

File: notebooks/cookbook/Querying and Modifying Charges and Other Parameters.ipynb [hunk 7]
```diff
@@ -25,6 +25,28 @@
...
+   "cell_type": "code",
+   "execution_count": 11,
+   "id": "552481f3",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "integrator = LangevinMiddleIntegrator(\n",
+    "    300 * unit.kelvin, 1 / unit.picosecond, 0.004 * unit.picoseconds)\n",
+    "simulation = Simulation(pdb.topology, system, integrator)\n",
+    "simulation.context.setPositions(pdb.positions)"
```
Comment (peastman) [inline notebooks/cookbook/Querying and Modifying Charges and Other Parameters.ipynb:47]: Why did you add this?  It's extra code that isn't relevant to subject.

File: notebooks/cookbook/Querying and Modifying Charges and Other Parameters.ipynb [hunk 8]
```diff
@@ -136,7 +149,10 @@
     "for i in range(bonded.getNumBonds()):\n",
     "    particle1, particle2, length, k = bonded.getBondParameters(i)\n",
     "    if particle1 == 1 and particle2 == 0:\n",
-    "        bonded.setBondParameters(i, particle1, particle2, length, k_)"
+    "        bonded.setBondParameters(\n",
+    "            i, particle1, particle2, length, \n",
+    "            2666 * unit.kilojoules_per_mole/unit.nanometer**2\n",
+    "        )"
```
Comment (peastman) [inline notebooks/cookbook/Querying and Modifying Charges and Other Parameters.ipynb:155]: No need to split this onto four lines.  It makes the code less readable, and it can easily fit on a single line.

File: notebooks/cookbook/Querying and Modifying Charges and Other Parameters.ipynb (hunk 7 above)
Comment (yuanqing-wang) [reply notebooks/cookbook/Querying and Modifying Charges and Other Parameters.ipynb:47]: I thought you would want to show how to modify the `Context`?

File: notebooks/cookbook/Querying and Modifying Charges and Other Parameters.ipynb (hunk 8 above)
Comment (yuanqing-wang) [reply notebooks/cookbook/Querying and Modifying Charges and Other Parameters.ipynb:155]: Sure! If we don't worry about the 80 char limit.

File: notebooks/cookbook/Querying and Modifying Charges and Other Parameters.ipynb (hunk 7 above)
Comment (yuanqing-wang) [reply notebooks/cookbook/Querying and Modifying Charges and Other Parameters.ipynb:47]: Should we just mention it in the text then?

**Comment by yuanqing-wang** [comment]
I've updated the PR accordingly.

File: notebooks/cookbook/Querying and Modifying Charges and Other Parameters.ipynb (hunk 5 above)
Comment (peastman) [reply notebooks/cookbook/Querying and Modifying Charges and Other Parameters.ipynb:159]: You marked this resolved, but you didn't remove the unnecessary line?

File: notebooks/cookbook/Querying and Modifying Charges and Other Parameters.ipynb (hunk 8 above)
Comment (peastman) [reply notebooks/cookbook/Querying and Modifying Charges and Other Parameters.ipynb]: If someone wants to view the entry on a VT52 terminal, I admit it will be inconvenient.  But I think we can say [VT100](https://en.wikipedia.org/wiki/VT100) is the earliest terminal we support. :)

**Comment by peastman** [comment]
Just one unnecessary line that still needs to be removed.  Then I think it will be ready to merge.

**Comment by yuanqing-wang** [comment]
> Just one unnecessary line that still needs to be removed. Then I think it will be ready to merge.

Sorry which unnecessary line?

**Comment by peastman** [comment]
```python
bonded = [f for f in system.getForces() if isinstance(f, HarmonicBondForce)][0]
```

**Comment by peastman** [comment]
Looks good.  Thanks!
